zenith_utils/src/postgres_backend.rs

```diff
@@ -450,7 +450,19 @@ impl PostgresBackend {
             }
 
             FeMessage::Execute(_) => {
-                handler.process_query(self, unnamed_query_string.clone())?;
+                trace!("got execute {:?}", unnamed_query_string);
+                // xxx distinguish fatal and recoverable errors?
+                if let Err(e) = handler.process_query(self, unnamed_query_string.clone()) {
+                    let errmsg = format!("{}", e);
+
+                    warn!(
+                        "query handler for {:?} failed: {:#}",
+                        unnamed_query_string, e
+                    );
+                    self.write_message(&BeMessage::ErrorResponse(errmsg))?;
+                }
+                // NOTE there is no ReadyForQuery message, this handler is used for basebackup.
+                // It uses copy protocol which expects ReadyForUqery after Sync message.
```

Comment:
I don't get this comment. 
`Sync` only present in Copy-both mode and basebackip is performed via `copy_out` (COPY TO STDOUT) protocol, which is very simple:

> Copy-out mode (data transfer from the server) is initiated when the backend executes a COPY TO STDOUT SQL statement. The backend sends a CopyOutResponse message to the frontend, followed by zero or more CopyData messages (always one per row), followed by CopyDone. The backend then reverts to the command-processing mode it was in before the COPY started, and sends CommandComplete. The frontend cannot abort the transfer (except by closing the connection or issuing a Cancel request), but it can discard unwanted CopyData and CopyDone messages.
> 
> In the event of a backend-detected error during copy-out mode, the backend will issue an ErrorResponse message and revert to normal processing. The frontend should treat receipt of ErrorResponse as terminating the copy-out mode.

Reply:
This is my understanding of the following:

> If the COPY command was issued via an extended-query message, the backend will now discard frontend messages until a Sync message is received, then it will issue ReadyForQuery and return to normal processing. If the COPY command was issued in a simple Query message, the rest of that message is discarded and ReadyForQuery is issued.

Initially I've copied issuing ReadyForQuery message from Query message handler but it resulted in an error. When I removed ReadyForQuery it started to work normally. So as I see this is the difference between flows for simple query and extended query

Reply:
> Initially I've copied issuing ReadyForQuery message from Query message handler but it resulted in an error. When I removed ReadyForQuery it started to work normally. So as I see this is the difference between flows for simple query and extended query


This documentation you have quotted is about copy-in mode, while we're in copy-out mode (sending data from pageserver TO client). 

Indeed, copy-out mode doesn't need an extra `ReadyForQuery` message, because the backend just switches back to normal processing mode after sending `CopyDone` or `ErrorResponse`

Reply:
Oh, indeed, thanks for pointing this out! I've changed the comment